File: solutions/fsharp/beer-song/BeerSong.fs
module BeerSong

let generateLine (bottles: int) : string list =
    let s_before =
        match bottles with
        | 1 -> ""
        | _ -> "s"

    let bottles_before =
        match bottles with
        | 0 -> "No more"
        | _ -> string bottles

    let bottles_after =
        match bottles with
        | 1 -> "no more"
        | 0 -> "99"
        | _ -> string (bottles - 1)

    let refrain =
        match bottles with
        | 0 -> "Go to the store and buy some more"
        | 1 -> "Take it down and pass it around"
        | _ -> "Take one down and pass it around"

    let s_after =
        match bottles_after with
        | "1" -> ""
        | _ -> "s"

    [ $"{bottles_before} bottle{s_before} of beer on the wall, {bottles_before.ToLower()} bottle{s_before} of beer."
      $"{refrain}, {bottles_after} bottle{s_after} of beer on the wall." ]

let rec recite (bottles: int) (takeDown: int) : string list =
    match takeDown with
    | 1 -> generateLine bottles
    | _ -> List.append (List.append (generateLine bottles) [ "" ]) (recite (bottles - 1) (takeDown - 1))

File: solutions/elixir/beer-song/lib/beer_song.ex
defmodule BeerSong do
  @doc """
  Get a single verse of the beer song
  """
  @spec verse(integer) :: String.t()
  def verse(number) do
    """
    #{String.capitalize(zero?(number))} #{plural(number)} of beer on the wall, #{zero?(number)} #{plural(number)} of beer.
    #{exhortation(number)}
    """
  end

  defp zero?(0), do: "no more"
  defp zero?(number), do: Integer.to_string(number)

  defp plural(1), do: "bottle"
  defp plural(_), do: "bottles"

  defp exhortation(0), do: "Go to the store and buy some more, 99 bottles of beer on the wall."
  defp exhortation(number) do
    now = number - 1
    "Take #{it_or_one(number)} down and pass it around, #{zero?(now)} #{plural(now)} of beer on the wall."
  end

  defp it_or_one(1), do: "it"
  defp it_or_one(_), do: "one"

  @doc """
  Get the entire beer song for a given range of numbers of bottles.
  """
  @spec lyrics(Range.t()) :: String.t()
  def lyrics(range \\ 99..0) do
    range
    |> Enum.map(&verse/1)
    |> Enum.join("\n")
  end
end
